### ISSUE TITLE
RFC: Links and Block References OPML Extension

### ISSUE THREAD

**Issue by andrewshell** [andrewshell/Tools-for-Thought-OPML#1]
## Goals of the Links and Block References OPML Extension

- OPML already has the `link` and `include` types. (See [Inclusion](http://opml.org/spec2.opml#1629042832000))
- The new elements are intended for functionality seen in LogSeq or Roam Research where there are potentially multiple tokens within the text of a headline that needs to be replaced with a link or the text of another headline. For example double-bracket wiki-style links.

## LogSeq Prior Art

- Here is a link to an example daily note markdown file from LogSeq: [2021_12_09.md
](https://raw.githubusercontent.com/andrewshell/Tools-for-Thought-OPML/master/LogSeq-Example/journals/2021_12_09.md)
- The block reference `((61b212cb-86bc-43af-a80b-f0e8f2f033f6))` refers to the id specified in [Prompts.md](https://raw.githubusercontent.com/andrewshell/Tools-for-Thought-OPML/master/LogSeq-Example/pages/Prompts.md)
- One issue that I'm looking to solve with the Links and Block References OPML Extension is that when you read a file, you should be able to identify what file the link or block references refer to.  
- With the current LogSeq markdown file you'd need to read and index the entire graph because although there is a default file structure, it can be changed by the user. 
- Additionally, when reading `2021_12_09.md` you wouldn't know that the `((61b212cb-86bc-43af-a80b-f0e8f2f033f6))` block was in the `Prompts.md` file unless you had already parsed and indexed `Prompts.md`.
- LogSeq allows for many different ways to link to other pages:
  - `#developer`
  - `#[[software developer]]`
  - `[[software developer]]`
  - `developer:: [[software]]`
  - `[developer](https://blog.andrewshell.org)`
  - `[developer]([[Andrew]])`
- Mostly these mirror the behavior of Roam Research, however, the double colon syntax differs. In Roam, you can have any title with spaces prior to the double colon and it will be turned into a bold link to that page (displayed with a single colon). In LogSeq the text before the double colon cannot have spaces and it doesn't become a link, it's more of a way to add metadata to the page.

## Glossary Prior Art

- Drummer (and other tools created by Dave Winer) has a feature called the [glossary](http://daytona.scripting.com/?q=glossary&collection=scriptingnews) (see [glossary.opml](http://drummer.scripting.com/davewiner/glossary.opml)) that allows the writer to automatically substitute phrases in their writing with another phrase.

## The "tft" Namespace

- OPML supports extensions as long as they are defined in their own namespace. I'm proposing the prefix "tft" (Tools for Thought) and the URI "https://tft.opmlns.org/" which will be the home for the final spec.

## What is a `<tft:link>` element?

- An `<outline>` can have zero or more `<tft:link>` elements that represent links within the outline text. This does not function like OPML inclusion, it should work like a web browser and open the linked document.
- There are two required attributes `text` and `href` and an optional `alt` attribute.
- The `text` attribute represents the text in the outline text that needs to be turned into a link.
- The `href` attribute represents what document it's linking to. Typically this would be another OPML file.
    - The `href` attribute should behave in the same way as an `href` in an HTML `<a>` element regarding relative vs absolute URLs.
- The `alt` attribute represents the displayed text of the link if different from the `text` attribute.

### Example:

```
<outline text="Today I'm meeting with [[Andrew]] to talk about [[OPML]].">
    <tft:link text="[[Andrew]]" href="Andrew.opml" alt="Andrew" />
    <tft:link text="[[OPML]]" href="OPML.opml" alt="OPML" />
    </outline>
```

## What is a `<tft:ref>` element?
- An `<outline>` can have zero or more `<tft:ref>` elements that represent text replacement within the outline.
- There are two required attributes `text` and `href`.
- The `text` attribute represents the text in the outline text that needs to be replaced.
- The `href` attribute represents what the source of the new text.
    - The `href` attribute should behave in the same way as an `href` in an HTML `<a>` element regarding relative vs absolute URLs.
    - If `href` points to an OPML or HTML file with no fragment (text after a hashmark), then we use the contents of the page title in the head element. If no title is present, ignore this element.
    - If `href` points to an OPML file with a fragment, then we look for the first `<outline>` element with `tft:id` attribute that matches the fragment and use the contents of the `text` attribute.
    - If there is no matching `tft:id` check for an `id` attribute (without namespace) because, in practice, OPML isn't strict about extra attributes being namespaced.
    - If `href` points to an HTML file with a fragment, then look for the first element in the document with a matching `id` attribute and use the contents of the element.

### Example:

```
<outline text="08:30 ((61b212cb))">
    <tft:ref text="((61b212cb))" href="Prompts.opml#61b212cb" />
    </outline>
```

## What is a `tft:id` attribute?

- An `<outline>` can have zero or one `tft:id` attribute that is a string that uniquely represents this headline in this document.

### Example:

```
<outline text="What is my top priority today?" tft:id="61b212cb" />
```

**Comment by scripting** [comment]
Andrew -- some feedback and questions per your request.

We tried to do something like this with &lt;scriptingNews> format which predates RSS in 1997, but it proved easier to just let HTML do the work, and include the links in the text itself, so we abandoned this idea when we adopted RSS 0.91.

That's how I do it these days with:

1. HTML anchor tags.

2. glossary references. 

3. emoji shortcodes. 

4. In fact -- I use [[syntax]] too in Scripting News, and guess what -- we just pass it through in the text attribute.

I'd be curious to see how LogSeq encodes the links in the Markdown text they generate, which I understand is their native file format. That's important prior art and imho should be in this document.

**Comment by andrewshell** [comment]
> I'd be curious to see how LogSeq encodes the links in the Markdown text they generate, which I understand is their native file format. That's important prior art and imho should be in this document.

Good call. I've uploaded a [sample LogSeq graph](https://github.com/andrewshell/Tools-for-Thought-OPML/tree/master/LogSeq-Example) and added the "LogSeq Prior Art" section to the issue above.

> * HTML anchor tags.

I'm not opposed to this, but I want to be careful. Although we would most likely be linking to another document via HTTP, the text of the outline might not be HTML. 

> * glossary references.
> * emoji shortcodes.

This is good prior art and block references could be used to implement a glossary or emoji shortcodes. However, with block references, you're embedding individual blocks from other notes. In a single document, you might reference blocks from many different other notes. Some referenced blocks might be from your own graph, while others could be from other notes that other people have published.

For example, you might create a block reference to someone else's note and then add your own commentary. Down the road, we could use something like [webmention](https://www.w3.org/TR/webmention/) where cross-site backlinks could be tracked.

> * In fact -- I use [[syntax]] too in Scripting News, and guess what -- we just pass it through in the text attribute.

One issue is that there is no standard for how to turn a wiki link into a URL. The other issue is that there are many different ways to create a link on these platforms. For instance, if you're pulling content in from Twitter, you might want to make `@andrewshell` a link to a note on your site. Right now, you could do that with the combination markdown/wikilink syntax `[@andrewshell]([[Andrew]])` but I'd like a simple method that is not tied to a specific implementation. That way, different publishing tools could experiment and the clients wouldn't need to know anything about the details.
